### PR TITLE
ref(actix): Migrate Aggregator actor to new tokio runtime [INGEST-1645]

### PR DESCRIPTION
**Pull request by olksdr**
Let's convert Metric Aggregator actor to `tokio` runtime.

In this PR:
* Create `AggregatorService` which runs on `tokio` runtime
* Some cleanup of the old messages and handlers
* Make sure all the Rust tests are working

Some open questions:
* how to test the new service when the both old and new runtime is required - like handling the shutdown message from the. controller?
* currently, just to keep changes to minimum, I use following
  ```
   if let Some(registry) = REGISTRY.get() {
                  registry
                      .aggregator
                      .send(InsertMetrics::new(self.project_key, metrics))
              }
  ```
  to get the aggregator from the registry and then send the message to it. This must be done better - share the channels with dependants?

### CODE REVIEW

File: relay-metrics/src/aggregation.rs [hunk 1]
```diff
@@ -1176,24 +1178,18 @@ impl Message for FlushBuckets {
...
-            .cost_tracker
-            .totals_cost_exceeded(self.config.max_total_bucket_bytes)
-    }
-}
-
 enum AggregatorState {
     Running,
+    // NOTE: This isn't used at the moment, since this service does not handle the shutdown
+    // messages just yet.
+    #[allow(dead_code)]
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: You need to use `Controller::subscribe_v2` and use the watch channel: see https://github.com/getsentry/relay/blob/95f154409f236ecc8089a4b19534997625a21735/relay-server/src/actors/health_check.rs#L129-L135 for an example of how we're currently thinking of handling this.  (though honestly that whole system probably needs changing as part of the more structured shutdown, but for now this is probably the way to do this)

File: relay-metrics/src/aggregation.rs [hunk 2]
```diff
@@ -1357,6 +1353,59 @@ impl<T: Iterator<Item = Bucket>> Iterator for CappedBucketIter<T> {
 
 impl<T: Iterator<Item = Bucket>> FusedIterator for CappedBucketIter<T> {}
 
+/// Flush message which triggers the flush cycle for the [`AggregatorService`]
+struct Flush;
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: `#[derive(Debug)]` probably, maybe also clone and copy

File: relay-metrics/src/aggregation.rs [hunk 3]
```diff
@@ -1824,49 +1913,46 @@ impl fmt::Debug for Aggregator {
...
+            loop {
+                sender.send(Flush).await.ok();
+                // Sleep for duration definited in `FLUSH_INTERVAL` and trigger the flush again.
+                tokio::time::sleep(FLUSH_INTERVAL).await;
+            }
+        });
+        tokio::spawn(async move {
+            // XXX: this requires the old system and runtime, and cannot be started in new tokio
+            // runtime ???
+            // let mut shutdown = Controller::subscribe_v2().await;
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: not sure i follow this?  healthcheck is the new runtime and uses this.  i just tried out this code and it seems to work?

File: relay-metrics/src/aggregation.rs [hunk 4]
```diff
@@ -2720,49 +2773,53 @@ mod tests {
     #[test]
     fn test_flush_bucket() {
         relay_test::setup();
+
         let receiver = TestReceiver::default();
+        // Create the new tokio runtime where we start AggregatorService
+        let rt = tokio::runtime::Runtime::new().unwrap();
+        // Run in the old actix system for receiver
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: Do we really still need the old runtime for this test?  Can we not convert TestReceiver to tokio as well?

File: relay-metrics/src/aggregation.rs [hunk 5]
```diff
@@ -3098,36 +3138,36 @@ mod tests {
...
+             "unit": "millisecond",
+             "value": [1, 1, 1, 1],
+             "type": "d",
+             "timestamp": 1615889440,
+             "width": 10,
+             "tags": {
+                 "route": "user_index"
+             }
+           }
+         ]"#;
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: this is all pretty noisy, is rustfmt no longer happy with the previous formatting?

File: relay-metrics/src/aggregation.rs [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
-use std::time::{Duration, Instant};
+use std::{
+    cmp::max,
+    collections::{btree_map, hash_map::Entry, BTreeMap, BTreeSet, HashMap},
+    fmt,
+    hash::Hasher,
+    iter::{FromIterator, FusedIterator},
+    mem,
+    time::{Duration, Instant},
+};
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: :( can we not change this import style?

File: relay-metrics/src/aggregation.rs [hunk 7]
```diff
@@ -15,15 +16,16 @@ use fnv::FnvHasher;
 use serde::{Deserialize, Serialize};
 
 use relay_common::{MonotonicResult, ProjectKey, UnixTimestamp};
-use relay_system::{Controller, Shutdown};
+use relay_system::{AsyncResponse, FromMessage, Interface, NoResponse, Sender, Service};
+use tokio::sync::mpsc;
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: This should be in the block of 3rd party crates above

File: relay-metrics/src/aggregation.rs [hunk 8]
```diff
@@ -1,11 +1,12 @@
...
+    cmp::max,
+    collections::{btree_map, hash_map::Entry, BTreeMap, BTreeSet, HashMap},
+    fmt,
+    hash::Hasher,
+    iter::{FromIterator, FusedIterator},
+    mem,
+    time::{Duration, Instant},
+};
 
 use actix::prelude::*;
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: This should be removed, should it not?

File: relay-metrics/src/aggregation.rs [hunk 9]
```diff
@@ -1388,15 +1437,16 @@ impl<T: Iterator<Item = Bucket>> FusedIterator for CappedBucketIter<T> {}
 /// all elapsed buckets belonging to the same [`ProjectKey`] are flushed together.
 ///
 /// Receivers must implement a handler for the [`FlushBuckets`] message.
-pub struct Aggregator {
+pub struct AggregatorService {
     config: AggregatorConfig,
     buckets: HashMap<BucketKey, QueuedBucket>,
     receiver: Recipient<FlushBuckets>,
```
Comment (flub) [inline relay-metrics/src/aggregation.rs]: `Recipient` is still something actix, why is that still here?

**Review by flub — COMMENTED** [review]
It seems there's still an lot of actix code left in here?  Is that intentional?

File: relay-metrics/src/aggregation.rs (hunk 3 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs]: The code itself works as expected, but test does not work when I have to start 2 different services/actors in 2 different systems, and I get the following error:
```
thread 'tokio-runtime-worker' panicked at 'System is not running', /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.9/src/system.rs:116:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'aggregation::tests::test_flush_bucket' panicked at 'called `Result::unwrap()` on an `Err` value: SendError', relay-metrics/src/aggregation.rs:2801:79
test aggregation::tests::test_flush_bucket ... FAILED

```

File: relay-metrics/src/aggregation.rs (hunk 1 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs]: see the comment in https://github.com/getsentry/relay/pull/1508#discussion_r985744067

File: relay-metrics/src/aggregation.rs (hunk 4 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs]: We could, but in production we still have `project_cache` actor which uses old system and we still have to test it. 
Once we convert that actor to the new tokio, we can change all of this as well.

File: relay-metrics/src/aggregation.rs (hunk 5 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs]: so, that what `cargo fmt` did.

File: relay-metrics/src/aggregation.rs (hunk 6 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs]: I might went overboard with this imports here 😅  I can revert this back.

File: relay-metrics/src/aggregation.rs (hunk 9 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs]: This service still interacts with `project_cache` actor which runs on the old actix.  I wanted to keep these changes to the minimum and then when the `project_cache` is converted we can remove this `Recipient` and maybe replace it with `mpsc` channel? 

I can look into it and maybe introduce changes in this PR, if that's better for the progress of migration.

Comment (flub) [reply relay-metrics/src/aggregation.rs]: the approach has been to use compat senders during the transition, I think that makes the most sense here as well.  the code here will become easier to grasp if it is fully converted

File: relay-metrics/src/aggregation.rs (hunk 4 above)
Comment (flub) [reply relay-metrics/src/aggregation.rs]: sure, but these tests are to make sure the aggregation actor works i think, not to make sure that the callers/senders mess something up?

File: relay-metrics/src/aggregation.rs (hunk 3 above)
Comment (flub) [reply relay-metrics/src/aggregation.rs]: we can fix that though, we've seen similar things before.  for this reason we have a helper to create the tokio runtimes which also enables the actix system: it allows sending actix messages from tokio to actix.  so this sounds like accidentally not having started the tokio runtime with that utility?  was this in tests or in relay itself?

File: relay-server/src/actors/health_check.rs [hunk 10]
```diff
@@ -96,9 +96,15 @@ impl HealthCheckService {
                     return false;
                 }
 
-                compat::send(Aggregator::from_registry(), AcceptsMetrics)
-                    .await
-                    .unwrap_or(false)
+                // NOTE: usually this unwrap is hidden in the `from_registry` function on the
+                // specific service or actor. But since the `AggregatorService` is in the different
+                // crate we cannot easily implement that pattarn and for now we have to use
+                // REGISTRY dirrectly and get the aggregator service out of it.
```
Comment (jjbayer) [inline relay-server/src/actors/health_check.rs]:
```suggestion
                // specific service or actor. But since the `AggregatorService` is in the different
                // crate we cannot easily implement that pattern and for now we have to use
                // REGISTRY directly and get the aggregator service out of it.
```

File: relay-server/src/actors/project.rs [hunk 11]
```diff
@@ -601,7 +602,12 @@ impl Project {
     /// The buckets will be keyed underneath this project key.
     pub fn merge_buckets(&mut self, buckets: Vec<Bucket>) {
         if self.metrics_allowed() {
-            Aggregator::from_registry().do_send(MergeBuckets::new(self.project_key, buckets));
+            // Get the `AggregatorService` from the registry and send the message to it.
+            if let Some(registry) = REGISTRY.get() {
```
Comment (jjbayer) [inline relay-server/src/actors/project.rs]: Is there a reason we do not `unwrap()` here, like in other usages?

File: relay-server/src/actors/health_check.rs [hunk 12]
```diff
@@ -96,9 +96,15 @@ impl HealthCheckService {
...
+                let aggregator = REGISTRY
+                    .get()
+                    .map(|registry| registry.aggregator.clone())
+                    .unwrap();
```
Comment (jjbayer) [inline relay-server/src/actors/health_check.rs]: Even if we cannot use `Aggregator::from_registry` here, couldn't we use `Registry::aggregator()`?

File: relay-system/src/compat.rs [hunk 13]
```diff
@@ -57,3 +57,26 @@ where
...
+where
+    M: Message + Send + 'static,
+    M::Result: Send,
+{
+    let (tx, rx) = oneshot::channel();
+    let f = futures01::future::lazy(move || addr.send(msg))
+        .then(|res| tx.send(res))
+        .map_err(|_| ());
+    EXECUTOR.get().unwrap().spawn(f);
+    rx.await.map_err(|_| MailboxError::Closed)?
```
Comment (jjbayer) [inline relay-system/src/compat.rs]: Of what type is the error that we map to `MailboxError::Closed` here?

File: relay-metrics/src/aggregation.rs [hunk 14]
```diff
@@ -1357,6 +1324,160 @@ impl<T: Iterator<Item = Bucket>> Iterator for CappedBucketIter<T> {
...
+///   failed buckets. They will be merged back into the aggregator and flushed at a later time.
+#[derive(Clone, Debug)]
+pub struct FlushBuckets {
+    /// The project key.
+    pub project_key: ProjectKey,
+    /// The logical partition to send this batch to.
+    pub partition_key: Option<u64>,
+    /// The buckets to be flushed.
+    pub buckets: Vec<Bucket>,
+}
```
Comment (jjbayer) [inline relay-metrics/src/aggregation.rs:1349]: It's a good idea to reorder these declarations, but it also makes the diff hard to read. What if we keep them where they were, and do the reordering in a separate PR that does not change anything functionally?

**Review by jjbayer — COMMENTED** [review]
Looks good to me in general, just some questions for my own understanding.

File: relay-server/src/actors/health_check.rs (hunk 12 above)
Comment (olksdr) [reply relay-server/src/actors/health_check.rs]: yeah, thanks for point this out, I've just missed few places where I had to replace this with a new function.

File: relay-system/src/compat.rs (hunk 13 above)
Comment (olksdr) [reply relay-system/src/compat.rs]: The function is exactly the same as in the `send` case above, but for `Recipient`. I can check what the error is in there as well

File: relay-metrics/src/aggregation.rs (hunk 14 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs:1349]: If this is difficult to parse I can move those structs back.  I decided to move all in one place so it either to see how the API of this service looks like and one can see all the messages and `impl`s for them in one place.

File: relay-server/src/actors/project.rs [hunk 15]
```diff
@@ -17,7 +17,7 @@ use relay_filter::{matches_any_origin, FiltersConfig};
 use relay_general::pii::{DataScrubbingConfig, PiiConfig};
 use relay_general::store::{BreakdownsConfig, MeasurementsConfig};
 use relay_general::types::SpanAttribute;
-use relay_metrics::{self, Aggregator, Bucket, InsertMetrics, MergeBuckets, Metric};
+use relay_metrics::{self, Bucket, InsertMetrics, MergeBuckets, Metric};
```
Comment (jan-auer) [inline relay-server/src/actors/project.rs]: While we're at it, this `self` can be removed.

```suggestion
use relay_metrics::{Bucket, InsertMetrics, MergeBuckets, Metric};
```

File: relay-server/src/service.rs [hunk 16]
```diff
@@ -194,8 +206,9 @@ impl ServiceState {
         let health_check = HealthCheckService::new(config.clone()).start();
         let relay_cache = RelayCacheService::new(config.clone()).start();
 
-        let aggregator = Aggregator::new(config.aggregator_config(), project_cache.recipient());
-        registry.set(Arbiter::start(|_| aggregator));
+        let aggregator =
+            AggregatorService::new(config.aggregator_config(), project_cache.recipient());
+        let aggregator = aggregator.start();
```
Comment (jan-auer) [inline relay-server/src/service.rs]: Please start the aggregator in a dedicated runtime for the time being. It is resource-intensive, mostly through allocations. Until we update its internals to use `spawn_blocking`, it is safer to follow the current thread allocation and have it in a dedicated, single-threaded runtime. 

See `outcome_runtime` for an example.

File: relay-system/src/compat.rs (hunk 13 above)
Comment (jan-auer) [reply relay-system/src/compat.rs]: The type of error is `tokio::sync::oneshot::error::RecvError` from the receiver's [`Future` implementation](https://docs.rs/tokio/latest/tokio/sync/oneshot/struct.Receiver.html#associatedtype.Output). Semantics are compatible for this case.

File: relay-system/src/controller.rs [hunk 17]
```diff
@@ -175,7 +175,6 @@ impl Controller {
 impl Default for Controller {
     fn default() -> Self {
         let (shutdown_sender, shutdown_receiver) = watch::channel(None);
-
```
Comment (jan-auer) [inline relay-system/src/controller.rs]: This can be restored.

File: relay-metrics/src/aggregation.rs [hunk 18]
```diff
@@ -641,7 +645,7 @@ fn tags_cost(tags: &BTreeMap<String, String>) -> usize {
 #[fail(display = "failed to parse metric bucket")]
 pub struct ParseBucketError(#[cause] serde_json::Error);
 
-/// An aggregation of metric values by the [`Aggregator`].
+/// An aggregation of metric values by the [`AggregatorService`].
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: I would keep linking to the `Aggregator` interface in this and similar doc comments (Bucket etc), since the interface describes the opaque behavior for metrics aggregation.

File: relay-metrics/src/aggregation.rs [hunk 19]
```diff
@@ -1357,6 +1324,160 @@ impl<T: Iterator<Item = Bucket>> Iterator for CappedBucketIter<T> {
...
+pub struct FlushBuckets {
+    /// The project key.
+    pub project_key: ProjectKey,
+    /// The logical partition to send this batch to.
+    pub partition_key: Option<u64>,
+    /// The buckets to be flushed.
+    pub buckets: Vec<Bucket>,
+}
+
+// NOTE(): required by ProjectCache and will be removed with ProjectCache migration.
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Let's mark this as `TODO`, which is easy to grep for. If we want to group those, we can even make it `TODO(actix)`, which is the scope used for the rewrite initiative.

File: relay-metrics/src/aggregation.rs [hunk 20]
```diff
@@ -1357,6 +1324,160 @@ impl<T: Iterator<Item = Bucket>> Iterator for CappedBucketIter<T> {
...
+
+/// Aggregator service interface
+#[derive(Debug)]
+pub enum Aggregator {
+    /// The health check message which makes sure that the service can accept the requests now.
+    AcceptsMetrics(AcceptsMetrics, Sender<bool>),
+    /// Insert metrics.
+    InsertMetrics(InsertMetrics),
+    /// Merge the buckets.
+    MergeBuckets(MergeBuckets),
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs:1433]: The actix `Message` implementations return a `Result` with an aggregation error on failure. For the time being, we should not change that contract and have a `Sender<Result<(), AggregateMetricsError>>` for these two.

File: relay-metrics/src/aggregation.rs [hunk 21]
```diff
@@ -1357,6 +1324,160 @@ impl<T: Iterator<Item = Bucket>> Iterator for CappedBucketIter<T> {
...
+    }
+
+    /// Returns the list of the buckets in the current `MergeBuckets` message, consuming the
+    /// message itself.
+    pub fn buckets(self) -> Vec<Bucket> {
+        self.buckets
+    }
+}
+
+/// Aggregator service interface
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]:
```suggestion
/// Aggregator service interface.
```

We should move the doc comment from `AggregatorService` to the interface, though. Outside of this module and service construction, the interface is the main type that other modules will interface with.

File: relay-metrics/src/aggregation.rs [hunk 22]
```diff
@@ -1357,6 +1324,160 @@ impl<T: Iterator<Item = Bucket>> Iterator for CappedBucketIter<T> {
...
+}
+
+impl FromMessage<MergeBuckets> for Aggregator {
+    type Response = NoResponse;
+    fn from_message(message: MergeBuckets, _: ()) -> Self {
+        Self::MergeBuckets(message)
+    }
+}
+
+/// Represents the closure which accepts the `Recipient` and `FlushBuckets` stract and return the
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]:
```suggestion
/// Represents the closure which accepts the `Recipient` and `FlushBuckets` struct and returns the
```

File: relay-metrics/src/aggregation.rs [hunk 23]
```diff
@@ -1742,36 +1864,12 @@ impl Aggregator {
...
-        );
-    }
-
     /// Sends the [`FlushBuckets`] message to the receiver in the fire and forget fashion. It is up
     /// to the receiver to send the [`MergeBuckets`] message back if buckets could not be flushed
     /// and we require another re-try.
     ///
     /// If `force` is true, flush all buckets unconditionally and do not attempt to merge back.
-    fn try_flush(&mut self) {
+    async fn try_flush(&mut self, sender: CompatSender) {
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Instead of passing the `sender` as parameter, can you access `self.recipient` and use `compat::send_to_recipient` directly? Particularly, please also note my comment below about `do_send`.

File: relay-metrics/src/aggregation.rs [hunk 24]
```diff
@@ -1791,13 +1889,22 @@ impl Aggregator {
...
                     let batch_size = batch.len() as u64;
-                    let result = self.receiver.do_send(FlushBuckets {
+                    let receiver = self.receiver.clone();
+                    let flush = FlushBuckets {
                         project_key,
                         partition_key,
                         buckets: batch,
-                    });
+                    };
+                    let result = sender(receiver, flush).await;
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Note that this changes semantics of the code:

- The old version used `do_send`, which sends a fire-and-forget message.
- The new version waits for the response of `FlushBuckets`. 

Let's not change semantics of the code and use the synchronous `.send()` with a `NoResponse` on `FlushBuckets`. If that requires us to introduce a second `FlushBuckets` message elsewhere, let's rather do that. 

This will also allow you to restore `fn process_batches` above.

File: relay-metrics/src/aggregation.rs [hunk 25]
```diff
@@ -1806,15 +1913,60 @@ impl Aggregator {
...
     }
+
+    async fn handle_message(&mut self, msg: Aggregator) {
+        match msg {
+            // This is the `HealthCheck` request where the `AsyncResponse` is expected.
+            Aggregator::AcceptsMetrics(AcceptsMetrics, sender) => {
+                let result = !self
+                    .cost_tracker
+                    .totals_cost_exceeded(self.config.max_total_bucket_bytes);
+                sender.send(result);
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Please introduce individual `handle_xyz` (e.g. `handle_merge_buckets`) methods and dispatch from `handle_message`. 

This is the pattern used in other services, too. We use this because it helps with navigating and reading code, especially when there are many messages.

Here is an example with both kinds of messages: https://github.com/getsentry/relay/blob/95f154409f236ecc8089a4b19534997625a21735/relay-server/src/actors/test_store.rs#L128-L133

File: relay-metrics/src/aggregation.rs [hunk 26]
```diff
@@ -1806,15 +1913,60 @@ impl Aggregator {
...
+                        relay_log::error!("failed to insert metrics: {}", err)
+                    }
+                }
+            }
+            Aggregator::MergeBuckets(MergeBuckets {
+                project_key,
+                buckets,
+            }) => {
+                if let Err(err) = self.merge_all(project_key, buckets) {
+                    relay_log::error!("failed to merge buckets: {}", err)
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: (part 2) These errors were propagated to the sender instead of being logged here. Please restore old behavior -- we can follow up with a dedicated PR that changes control flow.

File: relay-metrics/src/aggregation.rs [hunk 27]
```diff
@@ -1824,49 +1976,33 @@ impl fmt::Debug for Aggregator {
...
+                    biased;
 
-impl SystemService for Aggregator {}
 
-impl Handler<Shutdown> for Aggregator {
-    type Result = Result<(), ()>;
+                    // NOTE: `compat::send_to_recipient` is required by ProjectCache and will be removed with ProjectCache migration.
+                    _ = ticker.tick() => self.try_flush(|recipient, flush_buckets| Box::pin(compat::send_to_recipient(recipient, flush_buckets))).await,
+                    Some(message) = rx.recv() => self.handle_message(message).await,
+                    _ = shutdown.changed() => self.handle_shutdown(&shutdown.borrow()),
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Borrow does not reset the `watch`, so you need to update it as well. Otherwise, this will fire repeatedly. 

```suggestion
                    _ = shutdown.changed() => self.handle_shutdown(&shutdown.borrow_and_update()),
```

File: relay-metrics/src/aggregation.rs [hunk 28]
```diff
@@ -1824,49 +1976,33 @@ impl fmt::Debug for Aggregator {
...
-    fn handle(&mut self, message: Shutdown, _context: &mut Self::Context) -> Self::Result {
-        if message.timeout.is_some() {
-            relay_log::trace!("Shutting down...");
-            self.state = AggregatorState::ShuttingDown;
-        }
-        Ok(())
+                    else => break,
+                }
+            }
+            relay_log::info!("AggregatorService shutted down.");
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: The old messages were "aggregator started" and "aggregator stopped":

```suggestion
            relay_log::info!("aggregator stopped");
```

I'm happy with a new message, however see spelling:
```suggestion
            relay_log::info!("AggregatorService shut down.");
```

File: relay-metrics/src/aggregation.rs [hunk 29]
```diff
@@ -2007,6 +2045,14 @@ mod tests {
         fn bucket_count(&self) -> usize {
             self.data.read().unwrap().buckets.len()
         }
+
+        async fn bucket_count_async(&self) -> usize {
+            // We have to wait for a bit to let the test receiver to be called once to receive the
+            // message.
+            relay_test::block_fn(|| relay_test::delay(Duration::from_millis(10)).map_err(|_| ()))
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Do we still need this now that we have a utility that runs the actix system alongside tokio?

File: relay-metrics/src/aggregation.rs [hunk 30]
```diff
@@ -2128,68 +2174,68 @@ mod tests {
...
-            "value": [36, 49, 57, 68],
-            "type": "d",
-            "timestamp": 1615889440,
-            "width": 10,
-            "tags": {
-                "route": "user_index"
-            }
-          }
-        ]"#;
+           {
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Please revert indentation changes. These indents are now 3 + 2 spaces, which is inconsistent.

File: relay-metrics/src/aggregation.rs [hunk 31]
```diff
@@ -2386,8 +2432,8 @@ mod tests {
         assert_eq!(
             bucket_key.cost(),
             88 + // BucketKey
-            5 + // name
-            (5 + 5 + 6 + 2) // tags
+             5 + // name
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: Wrong indentation level.

File: relay-metrics/src/aggregation.rs [hunk 32]
```diff
@@ -2720,97 +2766,91 @@ mod tests {
...
-            let config = AggregatorConfig {
-                bucket_interval: 1,
-                initial_delay: 0,
-                debounce_delay: 0,
-                ..Default::default()
-            };
-            let recipient = receiver.clone().start().recipient();
-            let project_key = ProjectKey::parse("a94ae32be2584e0bbd7a4cbb95971fee").unwrap();
-            let aggregator = Aggregator::new(config, recipient).start();
+        let config = AggregatorConfig {
```
Comment (jan-auer) [inline relay-metrics/src/aggregation.rs]: From an offline conversation: If you move `config`, `recipient`, `project_key` into `rt.spawn(async move {})`, your diff will look much cleaner. 

Basically, all that has to change in this test is that you replace `relay_test::block_fn` with the new runtime, along with a channel.

I would propose to add a simple utility for this, similar to:

```rust
fn block_with_actix(f: impl Future<Output = ()>) {
    let system = actix::System::current();
    let (tx, rx) = tokio01::sync::oneshot::channel();
    
    let rt = tokio::runtime::Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    rt.spawn(async move {
        actix::System::set_current(system);
        compat::init();
        f.await;
        tx.send(()).ok();
    });

    // Note that this also resolves with error if `f` panics.
    relay_test::block_fn(|| rx).ok();
}
```

**Review by jan-auer — CHANGES_REQUESTED** [review]

File: relay-metrics/src/aggregation.rs (hunk 27 above)
Comment (flub) [reply relay-metrics/src/aggregation.rs]: the `shutdown.changed()` part makes sure that you see every value though.  And since we don't expect this to change a lot we shouldn't expect a new update that we definitely need to mark as seen in between .changed() returning and the borrow.  so i think the current version is clearer.

Comment (jan-auer) [reply relay-metrics/src/aggregation.rs]: You're entirely right, I had missed that `changed` also marks the value as seen. In that case, please disregard my comment - both ways are fine. 

Let's just make sure we use the same pattern everywhere.

File: relay-metrics/src/aggregation.rs (hunk 20 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs:1433]: As discussed, we never checked for the returned errors for these messages, so to simplify our interface we agreed on keeping `NoResponse`.

File: relay-metrics/src/aggregation.rs (hunk 26 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs]: Same as in https://github.com/getsentry/relay/pull/1508#discussion_r990594129 the errors in the sender were never logged or reported - we decided to report them here

File: relay-metrics/benches/aggregator.rs [hunk 33]
```diff
@@ -1,12 +1,11 @@
-use std::collections::BTreeMap;
-use std::fmt;
+use std::{collections::BTreeMap, fmt};
```
Comment (flub) [inline relay-metrics/benches/aggregator.rs]: nitpicking, but what makes this stuff gratuitously change like this?  is it some tooling?  what's wrong with the existing style?

FWIW while it's not explicitly called out in our how to write rust the example suggests the existing style: https://www.notion.so/sentry/HOWTO-Code-Rust-at-Sentry-7b35f165b10b4492bb95ebe1471a9ada#4f8cc9ec744c472ebcc89791400d519c

File: relay-metrics/src/aggregation.rs [hunk 34]
```diff
@@ -1792,81 +1904,105 @@ impl Aggregator {
...
-                        relay_log::error!(
-                            "Failed to flush the buckets, dropping {batch_size} buckets: {err}."
-                        );
+                    if let Some(receiver) = self.receiver.clone() {
+                        let flush_buckets = FlushBuckets {
+                            project_key,
+                            partition_key,
+                            buckets: batch,
+                        };
+                        compat::send_to_recipient(receiver, flush_buckets);
```
Comment (flub) [inline relay-metrics/src/aggregation.rs:1913]: cool, i agree this is the right way to do this :)

File: relay-metrics/src/aggregation.rs [hunk 35]
```diff
@@ -1792,81 +1904,105 @@ impl Aggregator {
...
+            Aggregator::MergeBuckets(msg) => self.handle_merge_buckets(msg),
+        }
     }
 
-    fn stopped(&mut self, _ctx: &mut Self::Context) {
-        relay_log::info!("aggregator stopped");
+    fn handle_shutdown(&mut self, message: &Option<Shutdown>) {
+        if let Some(message) = message {
+            if message.timeout.is_some() {
+                self.state = AggregatorState::ShuttingDown;
```
Comment (flub) [inline relay-metrics/src/aggregation.rs:1965]: there used to be a log message here, not sure if it's important.

File: relay-metrics/src/aggregation.rs [hunk 36]
```diff
@@ -1792,81 +1904,105 @@ impl Aggregator {
...
+                    Some(message) = rx.recv() => self.handle_message(message),
+                    _ = shutdown.changed() => self.handle_shutdown(&shutdown.borrow_and_update()),
 
-    fn handle(&mut self, message: Shutdown, _context: &mut Self::Context) -> Self::Result {
-        if message.timeout.is_some() {
-            relay_log::trace!("Shutting down...");
-            self.state = AggregatorState::ShuttingDown;
-        }
-        Ok(())
+                    else => break,
```
Comment (flub) [inline relay-metrics/src/aggregation.rs:2000]: The shutdown logic of this loop eludes me, when does the loop ever break?  IIUC the select will only execute the else branch if all other branches are disabled.  To disable a branch it needs to have a precondition that evaluates to false, but none of the branches have a precondition.  So this loop never finishes?  I may be having this wrong (and if I have it wrong because it's not very obvious rather than I'm just being dumb maybe it's worth explaining it in a comment)

File: relay-metrics/benches/aggregator.rs (hunk 33 above)
Comment (olksdr) [reply relay-metrics/benches/aggregator.rs]: This is basically VIM setup I did to merge all the imports if they are coming from the same crate. I will revert it to the previous state.

File: relay-metrics/src/aggregation.rs (hunk 35 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs:1965]: We already have another message when the aggregator will stop, I don't see the reason to add more logs here.

File: relay-metrics/src/aggregation.rs (hunk 36 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs:2000]: So the new behaviour mirrors the old one for time being. This loop will exit only when the runtime stops - this how it was in the actix system. 

Once the rest of the actors migrate to the new tokio, we will refactor the shutdown process.

File: relay-metrics/benches/aggregator.rs (hunk 33 above)
Comment (flub) [reply relay-metrics/benches/aggregator.rs]: Oh, is this vim-specific or some rustfmt or rust-analyzer setting that I don't know about?  How does this work?  I'm curious :)

File: relay-metrics/src/aggregation.rs (hunk 36 above)
Comment (flub) [reply relay-metrics/src/aggregation.rs:2000]: Fair enough, mind adding a comment clarifying that this currently doesn't ever exit the loop?  Because I was wondering if I misunderstood this.

File: relay-metrics/benches/aggregator.rs (hunk 33 above)
Comment (olksdr) [reply relay-metrics/benches/aggregator.rs]: My config for `rustfmt` contains
```toml
imports_granularity = "Crate"
```

and then in vim I run following when I want to auto format the file:
```bash
> rustfmt +nightly  <path to file>
```

File: relay-metrics/src/aggregation.rs (hunk 36 above)
Comment (olksdr) [reply relay-metrics/src/aggregation.rs:2000]: yeah, will do 👍

Comment (olksdr) [reply relay-metrics/src/aggregation.rs:2000]: done in [9c2fb4f](https://github.com/getsentry/relay/pull/1508/commits/9c2fb4fd17801b0232ee48db13ea0d0ca8892df8)

Comment (olksdr) [reply relay-metrics/src/aggregation.rs:2000]: done in [9c2fb4f](https://github.com/getsentry/relay/pull/1508/commits/9c2fb4fd17801b0232ee48db13ea0d0ca8892df8)

**Review by jan-auer — APPROVED** [review]
Looks good, thank you! 

In tests there are still some broken indentation fixes, if you look at the diff including whitespace you will see them. It would be great to get those reverted before merging this.

File: relay-metrics/src/aggregation.rs (hunk 36 above)
Comment (flub) [reply relay-metrics/src/aggregation.rs:2000]: thanks!

**Review by flub — APPROVED** [review]
same comment on all the indentation stuff - would be great to not pollute the commit history with noise like that

File: relay-metrics/benches/aggregator.rs (hunk 33 above)
Comment (jan-auer) [reply relay-metrics/benches/aggregator.rs]: In VSCode / rust-analyzer, I use this:

```json
{
  "rust-analyzer.imports.granularity.group": "module",
  "rust-analyzer.imports.prefix": "crate"
}
```

Will probably check that in. Unfortunately, that doesn't create separate a import group for all `relay_*` crates, but we can consider removing that convention.

Generally, we run with rustfmt defaults and no overrides.
